File: lib/05.js
function getSeatInfo(contents) {
	return contents.split(/\n/g).map((ticket) => {
		const code = ticket.replace(/F|L/g, 0).replace(/B|R/g, 1);
		const x = parseInt(code.slice(0, 7), 2);
		const y = parseInt(code.slice(7), 2);
		const seatId = x * 8 + y;
		return { x, y, seatId };
	});
}

module.exports = {
	A05(contents) {
		const tickets = getSeatInfo(contents);

		return Math.max(...tickets.map((ticket) => ticket.seatId));
	},

	B05(contents) {
		const tickets = getSeatInfo(contents);
		const ids = tickets.map((ticket) => ticket.seatId).sort((a, b) => a - b);

		for (let i = 1; i < ids.length; i++) {
			if (ids[i] - ids[i - 1] === 2) {
				return ids[i] - 1;
			}
		}
	},
};

File: lib/05.dart
class Seat {
  int x;
  int y;
  int seatId;
}

List<Seat> getSeatInfo(String contents) {
	return contents.split('\n').map((ticket) {
    String code = ticket.replaceAll(RegExp(r'F|L'), '0').replaceAll(RegExp(r'B|R'), '1');
    Seat seat = new Seat();
    seat.x = int.parse(code.substring(0, 7), radix: 2);
    seat.y = int.parse(code.substring(7), radix: 2);
    seat.seatId = seat.x * 8 + seat.y;
    return seat;
	}).toList();
}

int A05(String contents) {
  List<Seat> tickets = getSeatInfo(contents);
  List<int> ids = tickets.map((e) => e.seatId).toList();
  ids.sort((a, b) => a.compareTo(b));

  return ids.last;
}

int B05(String contents) {
  List<Seat> tickets = getSeatInfo(contents);
  List<int> ids = tickets.map((e) => e.seatId).toList();
  ids.sort((a, b) => a.compareTo(b));

  for (int i = 1; i < ids.length; i++) {
    if (ids.elementAt(i) - ids.elementAt(i - 1) == 2) {
      return ids.elementAt(i) - 1;
    }
  }
}
